File: frontend/src/types/budget.ts
import { ExpenseCategory } from './transaction';

export type BudgetStatus = 'on_track' | 'warning' | 'over';

export interface Budget {
  id: number;
  category: ExpenseCategory | string;
  limit_amount: number;
  period: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface BudgetCreate {
  category: ExpenseCategory | string;
  limit_amount: number;
  period?: string;
}

export interface BudgetUpdate {
  limit_amount?: number;
  is_active?: boolean;
}

export interface BudgetProgress {
  category: ExpenseCategory | string;
  limit_amount: number;
  spent: number;
  remaining: number;
  percentage: number;
  status: BudgetStatus;
  month: string;
}

export interface BudgetSuggestion {
  category: ExpenseCategory | string;
  suggested_limit: number;
  percentile: number;
  months_analyzed: number;
  monthly_history: number[];
}

File: backend/app/schemas/budget.py
from pydantic import BaseModel, validator
from typing import List, Optional
from datetime import datetime
from ..models.transaction import ExpenseCategory


class BudgetBase(BaseModel):
    category: str
    limit_amount: float
    period: str = "monthly"

    @validator('category')
    def validate_category(cls, v):
        valid_categories = {c.value for c in ExpenseCategory}
        if v not in valid_categories:
            raise ValueError(f"Invalid category: {v}")
        return v

    @validator('limit_amount')
    def validate_limit_amount(cls, v):
        if v <= 0:
            raise ValueError("limit_amount must be greater than 0")
        return v


class BudgetCreate(BudgetBase):
    pass


class BudgetUpdate(BaseModel):
    limit_amount: Optional[float] = None
    is_active: Optional[bool] = None

    @validator('limit_amount')
    def validate_limit_amount(cls, v):
        if v is not None and v <= 0:
            raise ValueError("limit_amount must be greater than 0")
        return v


class Budget(BudgetBase):
    id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


class BudgetProgress(BaseModel):
    category: str
    limit_amount: float
    spent: float
    remaining: float
    percentage: float
    status: str  # "on_track" | "warning" | "over"
    month: str  # "YYYY-MM"


class BudgetSuggestion(BaseModel):
    category: str
    suggested_limit: float
    percentile: float
    months_analyzed: int
    monthly_history: List[float]
